app/gui/view/graph-editor/src/component/node/input/area.rs

```diff
@@ -418,7 +418,7 @@ impl Area {
 
             width <- model.label.width.map(|t| t + 2.0 * TEXT_OFFSET);
             frp.output.source.width      <+ width;
-            frp.output.source.expression <+ model.label.content;
+            frp.output.source.expression <+ model.label.content.gate(&edit_mode);
```

Comment:
I don't like this solution. It makes `expression` output different from the `model.expression.code` if the node happens to be not edited right now. It may cause a host of problems if someone is not aware.

I will do the refactoring you mentioned in the description, especially because it is not a hard one: we could just fire `output.expression` in `init_new_expression` method (where we fire other signals as well).

Reply:
It's not that easy. 
First of all, we can't fire FRP outputs, can we? Outputs don't have the `emit` method, so you need an additional "private" input that would be connected to the output directly.
Secondly, we still need to provide the Presenter with the updates when the user edits the node. So `output.expression` needs to be fired in two places: in `init_new_expression` and in a similar way, on `model.label.content` changes. I will prepare a commit with an alternative implementation.

I guess the main issue is that `model.label` is used for two different activities.
1. Displaying the edited node expression
2. Displaying the `viz_code`
The current implementation relies on the fact that these two are always the same, but it is wrong. Here I filter out "not needed" updates, and I agree that it looks fragile.

Reply:
Indeed, I missed the fact that `init_new_expression` actually fires the ports _inputs_.

So maybe the most elegant solution is just to have `set_expression` input, and call `set_expression` method when input is fired, and return the new expression to the output (so `set_expression` method is no longer `pub(crate)`)

Reply:
I implemented this alternative solution by refactoring the `component::node::input::area` module. The refactoring might be hard to follow because huge chunks of code were moved to other places.